logstash/lib/logstash/outputs/loki.rb

```diff
@@ -0,0 +1,279 @@
+# encoding: utf-8
+require "logstash/outputs/base"
+require "logstash/namespace"
+require 'net/http'
+require 'concurrent-edge'
+require 'time'
+require 'uri'
+require 'json'
+
+# An example output that does nothing.
+class LogStash::Outputs::Loki < LogStash::Outputs::Base
+  require 'logstash/outputs/loki/batch'
+  require 'logstash/outputs/loki/entry'
+
+  config_name "loki"
+
+  ## 'A single instance of the Output will be shared among the pipeline worker threads'
+  concurrency :single
+
+  ## 'Loki URL'
+  config :url, :validate => :string,  :required => true
+
+  ## 'BasicAuth credentials'
+  config :username, :validate => :string, :required => false
+  config :password, :validate => :string, secret: true, :required => false
+
+  ## 'Client certificate'
+  config :cert, :validate => :path, :required => false
+  config :key, :validate => :path, :required => false
+
+  ## 'TLS'
+  config :ca_cert, :validate => :path, :required => false
+
+  ## 'Loki Tenant ID'
+  config :tenant_id, :validate => :string, :default => "fake", :required => false
+
+  ## 'Maximum batch size to accrue before pushing to loki. Defaults to 102400 bytes'
+  config :batch_size, :validate => :number, :default => 102400, :required => false
+
+  ## 'Interval in seconds to wait before pushing a batch of records to loki. Defaults to 1 second'
+  config :batch_wait, :validate => :number, :default => 1, :required => false
+
+  ## 'Array of label names to include in all logstreams'
+  config :include_labels, :validate => :array, :default => [], :required => true 
+
+  ## 'Array of label names to exclude from sending to loki'
+  config :exclude_labels, :validate => :array, :default => [], :required => false 
+
+  ## 'Extra labels to add to all log streams'
+  config :external_labels, :valide => :hash,  :default => {}, :required => false
+
+  ## 'Log line field to pick from logstash. Defaults to "message"'
+  config :message_field, :validate => :string, :default => "message", :required => false
+
+  public
+  def register
+    @uri = URI.parse(@url + '/loki/api/v1/push')
+    unless @uri.is_a?(URI::HTTP) || @uri.is_a?(URI::HTTPS)
+      raise LogStash::ConfigurationError, "url parameter must be valid HTTP, currently '#{@url}'"
+    end
+
+    @logger.info("Loki output plugin", :class => self.class.name)
+
+    @Channel = Concurrent::Channel
+    @entries = @Channel.new
+    @quit = @Channel.new(capacity: 1)
```

Comment:
I don't think you need capacity, meaning the sender will wait for the receiver to finish.

Reply:
will remove the capacity